### ISSUE TITLE
Dragon Ball Z Tenkaichi Tag Team disconnects during wireless battle

### ISSUE THREAD

**Issue by MajinDipu** [hrydgard/ppsspp#14361]
This thing right here. It’s been around here since dawn of civilization. 
![Screenshot_20210407_160819_org ppsspp ppsspp](https://user-images.githubusercontent.com/75547787/113849948-c71a4c00-97bb-11eb-870e-f3a2c97c9498.jpg)

I think anybody who tried to do multiplayer in DBZ TTT knows about this problem. During ad-hoc battles the game will display "Connection Lost" message and it will disconnect you from the match. It happens all the time. Can't even play this game peacefully with my friends.
The whole match is supposed to be played without any interruptions but this happens.
Doesn’t matter which device or hardware I use this is guranteed to happen. I twicked all the settings but it just won’t go away! Fast memory enabled or disabled it will happen.

Please fix this!

**Comment by MojoJojoDojo** [comment]
The game is like that its really sensitive to the distance and connection.
If you will try to play the game over Xlink Kai online you would also get disconnections even from the Ad-hoc menu lol.
The netcode is just not meant to be played online etc but if its locally over wifi then I dont know.

**Comment by MajinDipu** [comment]
> It does seem like the game worked better in the past though: https://www.youtube.com/watch?v=shTrcz-EVI8

> I wonder how they messed this game up because compared to Shin Budokai 2 that game got like a god hood netcode lol for a fighting game on PSP.

What version those guys were using? It was clearly ok back then

**Comment by ANR2ME** [comment]
Are you playing this between PC vs Android? if it is, it might be the same disconnection issue with Dissidia 012 when being played on 2 different architecture, there might be some inconsistencies on the data being sent (ie. floating point precision on a specific value)

**Comment by MajinDipu** [comment]
> Are you playing this between PC vs Android?
Right now yes, I am testing with PC vs Android bcz I only have one mobile. But in android vs android situation it also happens. I think disconnect happens less often in android vs android.

**Comment by MajinDipu** [comment]
So umm... no solution, eh?

**Comment by MojoJojoDojo** [comment]
Maybe try to disable the JIT or the VFPU option in the JIT.
You can do it in Tools -> JIT Or something.

**Comment by ANR2ME** [comment]
> 
> 
> So umm... no solution, eh?

Unlike crashes that can be easily caught where the issue was using debugger (as long they can be reproduced), disconnection in the middle of gameplay is hard to detects, since we don't have any clue which commit affecting it, and debugging on Android is a pain since the logcat often goes EOL and stopped printing anything.
The video didn't even mentioned which version was being used and whether all players are using the same platform or multiple platforms.

So it will probably takes a long time until we can findout what exactly happened.

**Comment by MojoJojoDojo** [comment]
> > So umm... no solution, eh?
> 
> Unlike crashes that can be easily caught where the issue was using debugger (as long they can be reproduced), disconnection in the middle of gameplay is hard to detects, since we don't have any clue which commit affecting it, and debugging on Android is a pain since the logcat often goes EOL and stopped printing anything.
> And the video didn't even mentioned which version was being used and whether all players are using the same platform or multiple platforms.
> 
> So it will probably takes a long time until we can findout what exactly happened.

I remember that the game disconnected before and you somehow fixed it on some commit but I dont remember which.
Maybe it got regressed.
Best to use 1.10.3 or earlier because I think it worked better there then.

**Comment by MajinDipu** [comment]
> I can try adding notification message on screen when detecting time that goes backward, so that we can at least findout if the disconnection occurred right after this notification or not.

OK I did not understand for the most parts but it was helpful. And that adding notification message thing might be a beginning step to fix this?

**Comment by MajinDipu** [comment]
> Maybe try to disable the JIT or the VFPU option in the JIT.
> You can do it in Tools -> Developer tools Or something.

Too lazy to try this right now but I will test it later :)

**Comment by MojoJojoDojo** [comment]
I think that disabling the VFPU or JIT helped me with getting Dissidia 012 to not disconnect...
But it makes the game very slow on some android devices.

**Comment by ANR2ME** [comment]
I wondered if the android build was compiled with something precise precision like on Visual Studio's `/fp:precise` or not...

**Comment by MajinDipu** [comment]
> I think that disabling the VFPU or JIT helped me with getting Dissidia 012 to not disconnect...
> But it makes the game very slow on some android devices.

OK can you tell me which option are you talking about specifically? There is JIT Debug tools in
Tools> Developer Tools> JIT Tools
And inside there are many other options. Inside JIT Tools everything is disabled including 6 different VFPU options.

Should I change CPU Core? It's set to Dynarec (JIT)

**Comment by unknownbrackets** [comment]
Hm.  Is it relying on both sides calculating floats to exactly the same bits?

-[Unknown]

**Comment by ANR2ME** [comment]
Not sure, but there is a possibility for the game to treat specific floating point value as a different value (ie. replacing/rounding) due to some issue, which might happen if the game was ported from a different platform.

**Comment by ANR2ME** [comment]
Btw, we're using `-mfpu=neon` on ARMv7 right?
It seems Neon is not IEEE754 compliant, according to this https://wiki.debian.org/ArmHardFloatPort/VfpComparison

**Comment by unknownbrackets** [comment]
We might also be doing fused calculations in some places, among other things.  We are not bit exact to PSP results in multiple ways, and platforms are not always identical to each other either.  I guess this may be a problem for networking...

Between PSP and PPSSPP though, you just have to look at i.e. #2990 to see differences in opinions on floating point calculation.  Much of the PSP VFPU was not IEEE754 compliant either.

-[Unknown]

**Comment by hrydgard** [comment]
NEON on ArmV7 is indeed not IEEE754 compliant, fortunately ARM64's NEON "ASIMD" is compliant.

As Unknown says, there can still be minor differences, the spec doesn't do a good job of defining signs of NaNs and stuff like that. It's a very difficult problem to solve.

**Comment by MajinDipu** [comment]
Hey guys, I have a info to share with you. Using IR Interpreter the game had a stable connection. Meaning disconnection happened rarely. But the con here is the game becomes super laggy with IR Interpreter. Dynarec JIT gives better speed but in multiplayer it gets disconnected almost immediately. 

So my stupid theory is that The laggier the game gets the more stable tge connection becomes. Although that's not a solution since laggy gamepl
is the worst gameplay

**Comment by ANR2ME** [comment]
Does it disconnects when both of you not doing anything (ie. not fighting)?
If it disconnects by just idling, both device (PC and Android) might not running at the same pace, causing the slower one to be flooded by packets from the faster one, thus the receive buffer gets full and new packets gets dropped since the slower one can't process the data fast enough, and because of this dropped packets it will eventually desync.

PS: And i don't think frame skipping will be helpful in this case, but switching who host and who join might help.

**Comment by MajinDipu** [comment]
OK I haven’t tested with idling but surely will. The thing is both device runs at virtually same speed. But I guess not running at same speed technically

**Comment by MojoJojoDojo** [comment]
On the video I shared before you could see the players overclock the emulated CPU to high levels before they play it.
So messing with the emulated CPU option in System can also help I guess.

**Comment by MajinDipu** [comment]
Update: Doing nothing in the match or simply idling doesn’t make the game disconnect. I did nothing for like 500 seconds and the game was fine. Then decided to fight a little and it got disconnected after minute.

**Comment by MajinDipu** [comment]
> On the video I shared before you could see the players overclock the emulated CPU to high levels before they play it.
> So messing with the emulated CPU option in System can also help I guess.

Thanks for the head up Mojo. I will check that out again

**Comment by MajinDipu** [comment]
So I tried with emulated clock speed to 1000 MHz and I think it helped to prevent us from getting disconnected big time. Needs more testing by other people on other devices. Also I changed the I/O timing method to "Fast" from "Host".

Only time it disconnected when something weird happened right before connection lost. My player Vegeta was charging up away from enemy Goku in my screen but on the other screen Vegeta was standing right in front of Goku. I hit one attack button and " Connection lost" massage showed up. What happened was Goku knocked me away and I tried to rush to him immediately pressing Up+O. I missed the timing on my android device but for some reason I was succesfully rush charged into Goku on PC.

Anyway I hope this info will be useful. But I still believe making the emulated CPU clock speed to 1000 MHz and I/O timing method to "Fast" actually helped a lot. But as I mentioned before this theory needs more testing by other people. Also wish there was some way to increase clock speed over 1000 MHz as 1000 is the max range.

**Comment by ANR2ME** [comment]
I think it's normal to get weird behavior when forcing the CPU to a certain clock, it could also causing the FPS to dropped on some games due to the extra burden from overclocking.

Did you overclock both side? have you tried overclocking only one side?

Btw, if both side showing different things on the screen and followed by disconnection, the game might be desynced, it may happen when packets being dropped/lost too many times.

**Comment by MajinDipu** [comment]
> Did you overclock both side? have you tried overclocking only one side?

Overclocked both sides. Maybe I will try overclocking one side only. Also how do you think it would affect the connection if we could increase the CPU clock speed over 1000 MHz (2000 MHz for example)?? Is it possible that the "Connection lost" error message will become an extremely rare occurrence if not gone forever by doing so?

**Comment by unknownbrackets** [comment]
Are you using "Force real clock sync"?  I would recommend "Host" io timing as well for network play, and try to avoid frameskip if possible.  The force 60 FPS setting might be good too.

-[Unknown]

**Comment by MajinDipu** [comment]
I tried with both Force Real Clock on a d off and results were almost the same. And frameskip is also on. What a weird problem

**Comment by MajinDipu** [comment]
Also this kind of problem was in GvG too right? Is it fixed in GvG? If yes then how?

**Comment by ANR2ME** [comment]
I don't think the issue where you get disconnected when getting hit by a certain attack while doing multiplayer Android vs PC ever got fixed

**Comment by MajinDipu** [comment]
oopsie

**Comment by MojoJojoDojo** [comment]
I noticed there is a theme with games that got this issue I think they are called  "Arena Fightiing/Brawler Fighting" games.
Games like: Dissidia 012 , DBZ Tag Team and the mentioned GvG get disconnected by attacking.
There is a Marvel game and a WWE that also fits this criteria but does not have the same error , instead it just desyncs but does not disconnect so you see the other players doing other stuff on the other screen lol.

**Comment by MajinDipu** [comment]
Funny enough a lot of people actually don’t get this error playing online. Beacuse in online the gameplay lags. I think lag actually prevents connection lost.

**Comment by ANR2ME** [comment]
> 
> 
> Funny enough a lot of people actually don’t get this error playing online. Beacuse in online the gameplay lags. I think lag actually prevents connection lost.

May be they happened to use the same platform, all windows or all android

**Comment by MojoJojoDojo** [comment]
> Funny enough a lot of people actually don’t get this error playing online. Beacuse in online the gameplay lags. I think lag actually prevents connection lost.

online its so super laggy its not even worth playing lol.

**Comment by MajinDipu** [comment]
> > Funny enough a lot of people actually don’t get this error playing online. Beacuse in online the gameplay lags. I think lag actually prevents connection lost.
> 
> May be they happened to use the same platform, all windows or all android

Using same platform helps not getting disconnected?

**Comment by ANR2ME** [comment]
> 
> 
> > > Funny enough a lot of people actually don’t get this error playing online. Beacuse in online the gameplay lags. I think lag actually prevents connection lost.
> > 
> > 
> > May be they happened to use the same platform, all windows or all android
> 
> Using same platform helps not getting disconnected?

Yes, if the issue is related to VFPU.

**Comment by ANR2ME** [comment]
@MajinDipu i've been trying to reproduce this without success, playing Android vs Windows PC.
I tested this on Free battle mode (no custom characters), 1vs1 (disabling 3rd & 4th CPU player) Goku vs Vegeta, i choose the first map, but i only use the rectangle and triangle button to attack.
Btw, i didn't overclock my CPU, i also tested with by both answering Yes and No when it asked me for compatibility with PSP-1000, but i didn't test using different answer on each player tho, since it doesn't make sense, so both players use the same answer.

PS: My android phone have a ARMv8 CPU with 2.3Ghz, so it doesn't struggle running this game full speed without frameskipping, if your phone couldn't get a stable 30 FPS during gameplay, it might cause stability issue.

**Comment by MajinDipu** [comment]
No my phone can easily run it at 30 FPS and at full speed. Did you ever try it Android vs Android tho? I can test it same PC with 2 PPSSPP window using localhost but I only have one phone. I just want to know if the problem occurs between 2 phones that can run it at full speed

**Comment by MajinDipu** [comment]
> @MajinDipu i've been trying to reproduce this without success, playing Android vs Windows PC.

OK English is not my first language. 😅
Did you mean you tried to recreate the problem and it didn’t disconnect for you?
Or did you mean you couldn’t succes on finding a solution and it always got disconnected for you?

**Comment by ANR2ME** [comment]
It never disconnect on my test.

**This is what i did, please try this too:**
1). Restore PPSSPP settings to default.
2). Make sure PSP Model is PSP-2000/3000 (apparently on Android the default is PSP-1000)
3). Enable WLAN, Enable built-in Adhoc Server, and setup your pro Adhoc Server IP address
4). Set Port Offset to 5000
5). Don't change anything else, and play the game.

**In-game:**
1). Select Multiplayer
2). Player1 Create Free battle mode, No custom characters
3). Player2 Join the room, and choose Team B
4). Disable Player3 & Player4 CPU
5). Select Goku and Vegeta
6). Select the first Map, and start the game
7). Attack each other, it never disconnects in the middle of battle on my test.

**Comment by MajinDipu** [comment]
Hmm, OK. So I keep android to PSP-1000?

**Comment by ANR2ME** [comment]
> 
> 
> Hmm, OK. So I keep android to PSP-1000?

No, use PSP-2000/3000 on both of them.

**Comment by MajinDipu** [comment]
Alright I will do it

**Comment by ANR2ME** [comment]
Btw, please use the latest version at https://buildbot.orphis.net/ppsspp/ for testing this.

**Comment by MajinDipu** [comment]
I am using one of the latest version of the orphis builds not exactly the build that was released an hour ago tho.

Anyway I tested multiplayer 2 windows on one PC and there was absolutely no issue at all. Then I did 1v1 Android vs PC and it was fine. We also had an intense shot exchange the game lagged a bit but it was fine. Then I did 2v1 (2 Human vs CPU) and connection lost occured. I also played 2v1 in PC vs PC but it was clearly fine. I need to test more 1v1 in Android vs PC. (I switched to OpenGL on android from Vulkan. Kinda feel like it helped connection lost happening later than earlier.)

**Comment by MajinDipu** [comment]
Now, did you test android vs android? I need to find my rich friend and 1v1 with his high end mobile. If it doesn't disconnect then I will be sure the problem is happening in Android vs PC for me. And if it still disconnects I will have to find 2 high end mobile and test. And if I see that the problem stopped occurring between high end phones then I gotta say, maybe my phone gameplay secretly slows down during multiplayer hence all this connection lost.

**Comment by ANR2ME** [comment]
> I am using one of the latest version of the orphis builds not exactly the build that was released an hour ago tho.
> 
> Anyway I tested multiplayer 2 windows on one PC and there was absolutely no issue at all. Then I did 1v1 Android vs PC and it was fine. We also had an intense shot exchange the game lagged a bit but it was fine. Then I did 2v1 (2 Human vs CPU) and connection lost occured. I also played 2v1 in PC vs PC but it was clearly fine. I need to test more 1v1 in Android vs PC. (I switched to OpenGL on android from Vulkan. Kinda feel like it helped connection lost happening later than earlier.)

From this, it seems it only disconnects when there are CPU players

**Comment by MojoJojoDojo** [comment]
So the disconnetion issue is very specific huh? sound weird.

**Comment by MajinDipu** [comment]
> > I am using one of the latest version of the orphis builds not exactly the build that was released an hour ago tho.
> > Anyway I tested multiplayer 2 windows on one PC and there was absolutely no issue at all. Then I did 1v1 Android vs PC and it was fine. We also had an intense shot exchange the game lagged a bit but it was fine. Then I did 2v1 (2 Human vs CPU) and connection lost occured. I also played 2v1 in PC vs PC but it was clearly fine. I need to test more 1v1 in Android vs PC. (I switched to OpenGL on android from Vulkan. Kinda feel like it helped connection lost happening later than earlier.)
> 
> From this statement, it seems it only disconnects when there are CPU players.
> Is this CPU Player using a different character other than Goku/Vegeta? Might be the disconnection only happened with a certain character.

I don't think it happens due to certain characters. But one thing for sure that it is guaranteed to happen with CPU character. And in 1v1 it’s like a lottery weather it will happen or not. Let's forget the CPUs. Why does it happen in 1v1 is the real question

**Comment by ANR2ME** [comment]
If you are only using Goku and Vegeta characters, did it ever disconnects?
When you are getting the disconnection, which characters are there in the game? and which map being selected? also how long is the count down timer being set?

I need to be able to reproduce the issue in order to find out hat happened, and need a more detailed step-by-step setup like the test i did that never gets disconnected on my earlier post above.

**Comment by MajinDipu** [comment]
Hmm listen. I am gonna get my buddy and have 10 competitive 1v1 matches. Spamming button alone with 2 device won't give me specific result. Each time we r gonna try different characters and different map with 240 sec countdown. Clock shouldn’t be an issus and I am 100% sure about it. If we get disconnected I am gonna note it down what move was delivered moments before it happened. I would record the matches if I can. If we get disconnected then we will play on that same map with same characters 3 more times again. We will also try to recreate the problem by doing exact same move. I am gonna go all out now! Hopefully I will find something important about this crap.

**Comment by MajinDipu** [comment]
Also I remember last night when I was doing Vegeta vs Goku everything was ok for a while. Then I hit Goku with a single ki attack (Which is done by pressing simply triangle) and it got disconnected as soon as it made contact. It is to be noted that both men were hitting each other with countless ki attacks before in that match but it was all good.

**Comment by ANR2ME** [comment]
Well out many possible numbers on a floating point, only a few of them (special values) could be inconsistent across different platform, so that hit that cause a disconnection might bumped into such values/numbers, might be happened during collision calculation.

I'll try to create a test build with `/fp:fast` instead of `/fp:precise` since it will handle special values differently (ie. if x = NaN, x - x = 0 on /fp:fast, while x - x = NaN on /fp:precise), since it's normal for games to be compiled with faster preference over precision.
https://devblogs.microsoft.com/cppblog/do-you-prefer-fast-or-precise/

**Comment by MajinDipu** [comment]
Even though I don’t understand coding I think I am starting to understand a lot. Good luck with that build. Who knows maybe you will solve a problem that happens in all Arena Fighting games on the PSP. Save this genre with your test build!! 🙏🏻😭

**Comment by MajinDipu** [comment]
> **Update:**
> You can use this Windows builds for testing: https://www.dropbox.com/s/2v0vw9hm5gv7se9/PPSSPP_1.11-testbuild-fpfast_Win32x64.zip?dl=0
> For Android you can use the latest one at https://buildbot.orphis.net/ppsspp/ (hopefully Android built using Neon, and hopefully Neon prioritize faster performance over precision accuracy)

Oh it’s here? Man you are fast. It's past my bedtime lol. I am gonna test with these two builds as soon as I wake up 😂😅

**Comment by ANR2ME** [comment]
I'll remove that previous testbuild, here is a new link where Windows build is compiled using `/fp:fast`
Android: https://www.dropbox.com/s/b41bm43mtn1gpnn/PPSSPP_1.11-testbuild_ARMv7.apk?dl=0
Windows: https://www.dropbox.com/s/2t3mtdhb0f045cn/PPSSPP_1.11-testbuild_Win32x64.zip?dl=0
This test build will shows a message saying  "Time is going backward: xxxx -> xxxx", just in case the disconnection were related to an issue where time can go backward on PPSSPP.

**Comment by MojoJojoDojo** [comment]
If this fixes the Dissidia issue @zakilj3 might be interested.

**Comment by ANR2ME** [comment]
The only disconnection i had on DBZ TTT was when the fight is over and the screen is showing Rematch menu, if i didn't cheese anything sometimes it shows disconnection message, even with multiple instance.

Here is the log on the Host side:
![image](https://user-images.githubusercontent.com/7974720/115238782-46a70400-a148-11eb-9042-fef70ab5be0a.png)

And this is the log on Joining side:
![image](https://user-images.githubusercontent.com/7974720/115238830-56bee380-a148-11eb-8dd5-af24b5e4d129.png)

Based on the log, it seems the host is sending data and expecting a reply within 33~34ms (probably because it's a 30 FPS  game), if it didn't get a reply it try to resend again, however near the disconnection (indicated by `sceNetAdhocPdpDelete`) the host is getting the reply after 68ms (which is more than 2 frames difference), and the game seems to try to re-send only once.

So the main issue is probably because sometimes the packet taking too long to arrived after the game no longer attempt to retry.
In this case, i'm not sure what kind of "trick" i should do to make sure there is always incoming data within a frame duration (even if it's a duplicate of previous packet, pretending a packet from a retry) in order to prevent disconnection, in creasing the rec buffer size might not work, since the game will try to recv again within 1ms until the buffer is empty (based on the log on Joining side)

**Comment by MajinDipu** [comment]
OK in our first test match using Goku vs Vegeta in the first map the game disconnected when Goku hit me with a ki attack. We were throwing Ki attacks before but nothing was wrong. It was like day before yesterday situation. Also your time is going backward message didn’t show up. However it showed up when I started the game again on android. Here is the screenshot. 
![Screenshot_20210419_193508_org ppsspp ppsspp](https://user-images.githubusercontent.com/75547787/115245415-b6b48a80-a146-11eb-8ab9-de16b930dce9.jpg)

It is also happening Everytime I am booting the game.

**Comment by MajinDipu** [comment]
> So the main issue is probably because sometimes the packet is taking too long to arrived after the game is no longer attempt to retry (ie. the Client is sending it at 08:829 but arrived on Host side at 08:862, and this was tested on localhost)
> In this case, i'm not sure what kind of "trick" i should do to make sure there is always incoming data within a frame duration (even if it's a duplicate of previous packet, pretending a packet from a retry) in order to prevent disconnection, increasing the rec buffer size might not work, since the game will try to recv again within 1ms until the buffer is empty (based on the log on Joining side)

Damn! You got it! At least you are the first person in the world to know what's the main problem here. So is it unfixable? A whole genre is depending on this question!

**Comment by ANR2ME** [comment]
> 
> 
> OK in our first test match using Goku vs Vegeta in the first map the game disconnected when Goku hit me with a ki attack. We were throwing Ki attacks before but nothing was wrong. It was like day before yesterday situation. Also your time is going backward message didn’t show up. However it showed up when I started the game again on android. Here is the screenshot.
> ![Screenshot_20210419_193508_org ppsspp ppsspp](https://user-images.githubusercontent.com/75547787/115245415-b6b48a80-a146-11eb-8ab9-de16b930dce9.jpg)
> 
> It is also happening Everytime I am booting the game.

Hmm.. that is strange to get the current time = 0 (ie. previous -> current), might be because the FPS/VPS is 0 too.

**Comment by MajinDipu** [comment]
Yeah maybe u r right

**Comment by MajinDipu** [comment]
OK I also saw the "Time is going backwards" message while waiting in the lobby alone for like 30 sec. I missed the screenshot but it was here.
![Uploading Screenshot_20210419_194326_org.ppsspp.ppsspp.jpg…]()

**Comment by ANR2ME** [comment]
Yeah, the time going backward issue is a bit random, but as long it doesn't cause disconnection, it should be fine.

**Comment by ANR2ME** [comment]
> 
> 
> > So the main issue is probably because sometimes the packet is taking too long to arrived after the game is no longer attempt to retry (ie. the Client is sending it at 08:829 but arrived on Host side at 08:862, and this was tested on localhost)
> > In this case, i'm not sure what kind of "trick" i should do to make sure there is always incoming data within a frame duration (even if it's a duplicate of previous packet, pretending a packet from a retry) in order to prevent disconnection, increasing the rec buffer size might not work, since the game will try to recv again within 1ms until the buffer is empty (based on the log on Joining side)
> 
> Damn! You got it! At least you are the first person in the world to know what's the main problem here. So is it unfixable? A whole genre is depending on this question!

It can only be fixed by a workaround, probably by duplicating previous packet if recv buffer is empty on every frame.
But i don't like workaround too much, and the game can also get confused if the packet keeps appearing on every frame.

**Comment by MajinDipu** [comment]
OK can you tell me what does this workaround mean? workaround in PPSSPP or the game file?

**Comment by MojoJojoDojo** [comment]
If you make the game output 60 fps all the time there wont be any issue no?
Maybe a cheat code that could allow that or a mode in the game.

**Comment by MajinDipu** [comment]
Hey Mojo, I also heard there's many 30 fps games 
that suffers in multiplayer. Adding a 60 FPS cheat helps. But the problem is there's not a option to make the game run at 60 FPS(Maybe it’s not possible to have a option like that?). Instead someone gotta make a specific cheat for a specific game to make sure it runs at 60 FPS all the time. So it is kinda hard. I wish I knew how to make a 60 FPS cheat.
Oh I didn’t search on the internet if a cheat like this exist for DBZ TTT.

**Comment by hrydgard** [comment]
4 days is nothing :)

**Comment by MajinDipu** [comment]
OK I was getting used to hourly replies :)

**Comment by ANR2ME** [comment]
There won't be any update anytime soon, since i can't test whether duplicating the packet going to work or not due to we currently doesn't manage/keep track socket's buffer.

But i was planning to create a PSPSocket class to be used for infrastructure mode and merging the socket with AdhocSocket, thus we can manage each socket buffer on the class. Probably after i'm done working with infrastructure API.

For the meantime, if you can test whether cheat for 60 FPS help preventing the disconnection issue (in case it will cause data sync to be done at a shorter interval, to make sure there will always be packet received within 33ms).

PS: I was planning to create the 60 FPS cheat myself, but i forgot which syscall was it that can be used to wait for VBlank😅 Anyone remember? The one where the arg is 1 for 60 FPS dan 2 for 30 FPS.

**Comment by unknownbrackets** [comment]
`sceDisplayWaitVblankStartMulti()` is the one.

-[Unknown]

**Comment by MajinDipu** [comment]
It's cool ANR2ME. Take your time.
And I searched the whole internet for the 60 FPS cheat code without any luck. Also asking people if they ever made any patch/cheat for this game to run at 60 FPS but again no luck :(

**Comment by ANR2ME** [comment]
Hmm.. looks like the reason why it didn't get the packet after 2 frames and deleting the connection, because it wasn't even trying to recv the packet on one of them frame rendering.. i wondered why it skipped doing any networking on that frame other than checking the WLAN state
![image](https://user-images.githubusercontent.com/7974720/115820402-ec998d80-a42a-11eb-8079-da8074999cd1.png)

**Comment by ANR2ME** [comment]
Hmm.. i'm not sure how to create the cheat properly.
This is the code (`li a0,0x2`) which set the FPS when multiplayer game started, where this `0x00000002` value will be stored at `0x08A78E0C` and being read as arg of `sceDisplayWaitVblankStartMultiCB`
![image](https://user-images.githubusercontent.com/7974720/115826252-cc6ecc00-a434-11eb-880c-124dc9d59d8f.png)
After multiplayer ended the game will set the value at `0x08A78E0C` to `1`

I tried to create the cheat like this:
```
_S ULUS-10537
_G DBZ TTT [US] 
_C1 60 FPS
_L 0x08AB7EC8 0x34040001
```
But when i checked the instruction and the value stored in memory, they're not changing :( did i use the address in the wrong way? (in case the cheat have a base address)

**Comment by MajinDipu** [comment]
> **Edit:** LOL it did have a count down timer during "Rematch menu" 😅 i didn't noticed it because i was using 1x Window and the timer is covered by FPS info
> So i guess the disconnection during "Rematch menu" is normal, and now we're back to square one... don't know why some attack could cause disconnection, since i can't reproduce it (yet)

Are you sure that the disconnection in match and after the match happens for different reasons? Yes I know it has a countdown timer but do you think that packet and floating point things might be same in 2 situations? Or just completely different?

**Comment by MajinDipu** [comment]
> **Update:** Looks like i need to subtract the PSP address with `0x08800000` and added it with `0x20000000`
> So the correct cheat is:
> 
> ```
> _S ULUS-10537
> _G DBZ TTT [US] 
> _C1 60 FPS
> _L 0x202B7EC8 0x34040001
> ```
> 
> But the animation seems to be moving too fast.. probably not a good idea to use 60 FPS other than for testing purpose.

Bruh, I guess it’s so complex to make a 60 FPS cheat for this game that everyone who tried failed. Indeed this is one weird game. Anyway I will use the cheat. For testing purposes only lol

**Comment by ANR2ME** [comment]
> 
> 
> > **Edit:** LOL it did have a count down timer during "Rematch menu" 😅 i didn't noticed it because i was using 1x Window and the timer is covered by FPS info
> > So i guess the disconnection during "Rematch menu" is normal, and now we're back to square one... don't know why some attack could cause disconnection, since i can't reproduce it (yet)
> 
> Are you sure that the disconnection in match and after the match happens for different reasons? Yes I know it has a countdown timer but do you think that packet and floating point things might be same in 2 situations? Or just completely different?

The disconnection after the match is normal, while the one during battle is not normal, i'm still suspecting it has something to do with collision calculation when an attack hit the target based on the symptom similarity to GvG Next Plus which was suspected to be related to VROT/VCOS.

Anyway, it's difficult to fix since i haven't been able to reproduce it my self during battle, thus unable to check the log.

**Comment by ANR2ME** [comment]
Btw, if it happened quite often on your test, could you test disabling some of the JIT feature on both players, like this:
![image](https://user-images.githubusercontent.com/7974720/115865660-d4456500-a462-11eb-909f-05299f0b37eb.png)
If it didn't disconnect during battle, try removing each section (SIMD, LSU, FPU, VFPU) on each test to findout whether it's related to JIT or not. (checklisted = disabled JIT)

**Comment by MajinDipu** [comment]
> Btw, if it happened quite often on your test, could you test disabling some of the JIT feature on both players, like this:
> ![image](https://user-images.githubusercontent.com/7974720/115865660-d4456500-a462-11eb-909f-05299f0b37eb.png)
> If it didn't disconnect during battle, try removing each section (SIMD, LSU, FPU, VFPU) on each test to findout whether it's related to JIT or not, if it ever get disconnected it might be related. (btw, checklisted = disabled JIT)

Will do. But everything is already unticked. I will enable everything then try disabling everything one by one.

**Comment by MojoJojoDojo** [comment]
There is that PR here that chagnes stuff related to VFPU : 
https://github.com/hrydgard/ppsspp/pull/14406
Maybe it helps with this issue if it is related to VFPU...

**Comment by MajinDipu** [comment]
It's merged. Does this mean if it’s related to VFPU and I use latest versions the problem might be fixed?

**Comment by MajinDipu** [comment]
@ANR2ME Hey man, I got a tips to reproduce that connection lost thing. It's not in 1v1 tho. And requires a CPU  But might help. Here's what you have to do.
   1) Setup a 2v1 battle. Have 2 human players in one team and 1 CPU on the other.
   2) Set CPU difficulty to very weak.
   3) Pick any characters and any stage
   4) When the match starts both human player should approach CPU immediately. Very much possible to do alone.
   5) When 2 human player are close to the enemy, start punching and kicking (square button) simultaneously. Don't do the punch punch smash away and teleport kick thing. Hit 2 punch then take a breath then 2 punch again. 2 human should do it simultaneously. 
   5) Another things that you can do are ki attacks together both from close range and far away. Or use your super attacks together like Kamehameha and Gallick Gun to the CPU at the same time. Have one player do the kamehameha while other one does dragon rush. etc

This has 100% success rate of disconnecting. Good luck.

**Comment by ANR2ME** [comment]
> 
> 
> It's merged. Does this mean if it’s related to VFPU and I use latest versions the problem might be fixed?

Have you tried with the latest version to see whether that PR fixed this issue?

**Comment by MajinDipu** [comment]
Oops. Ok I didn’t test it on the latest build. To answer your 2nd question disconnection can occur in step 5 or step 6. Step 6 is divided into two parts but it might happen simply with ki attacks. So I hope kamehameha or gallick gun is not needed.
I know it is hard to simultaneously control both android and PC but I said in step 4 it is very much possible if you set CPU's difficulty to very weak. At least you know how to punch and kick. That will be it.

**Comment by MajinDipu** [comment]
@ANR2ME Yoooo, you gotta check this out! The latest build made this game unstable in multiplayer. Everything is sooo fast and acting weird all of a sudden.
Here's the video link -> https://youtu.be/NCbTLKTJEuQ
Needs testing by others with this build before creating a new issue. Who knows it might be my device. Please anyone who sees this, u gotta try multiplayer with the latest build. In case you are wondering if it is happening with other games then no. Tested Tekken 6 and it was fine and just as before. Not sure about other Arena fighters like Dissidia or Gundam tho.

**Comment by unknownbrackets** [comment]
Is it fine in v1.11.2-550-gdbe665880 and bad in v1.11.2-558-g0ccc63b43?  That does look potentially like some kind of not good sin/cos problem...

-[Unknown]

**Comment by MajinDipu** [comment]
I think I had v1.11.2-543 before and it was good. Not sure about 550

**Comment by unknownbrackets** [comment]
Well, I'm surprised we didn't have a vrot test for `cos(2 * pi/2)`.  Sorry, will be fixed in a84df2536.

-[Unknown]

**Comment by MajinDipu** [comment]
GG

**Comment by ANR2ME** [comment]
So did the latest version right now fixed this disconnection issue?

**Comment by MajinDipu** [comment]
OK only tested 2 match. 1st match was the cleanest match we had in a long time. Everything went perfectly. Also we tried to do many things to force disconnection but it was all good. After the 1st match we pressed rematch. In the 2nd match again that disconnection occurred. Needs more testing but rn Imma sleep. I have this weird feeling that the connection might've actually improved. But that's too early to say bcz only played 2 match. But damn! That 1st match was magical!

**Comment by ANR2ME** [comment]
May be the game uses more VFPU other than VROT and VCOS (i'm not sure what kind of calculation usually used to checks for collision, but usually the colliders are in simple shape like box, sphere, and capsule).
But it's nice to know you had your first ever stable gameplay :D

**Comment by MojoJojoDojo** [comment]
Tried to find out what happens with Android to PC on Dissidia 012...
WIth the latest dev build it will disconnect not long after the match starts.
But if you disable the VFPU stuff from the JIT the match seems to last longer but will still disconnect.

**Comment by MojoJojoDojo** [comment]
> I don't think the issue where you get disconnected when getting hit by a certain attack while doing multiplayer Android vs PC ever got fixed on GvG.
> 
> It was an old bug, i think it was first reported on forum, but between JPCSP vs PPSSPP https://forums.ppsspp.org/showthread.php?tid=12357&pid=99304#pid99304
> 
> > I've noticed that the error (2:45) happens only after someone receives an attack, specially the triangle button attacks (or a couple of seconds after). If the attacks don't hit, the error won't show up.
> > As the battle gets more and more chaotic the error will pop up, if any AI is present the error will happen a lot more faster (5 to 30 seconds).
> > Of course, this error only happens when using JPCSP as host and PPSSPP as player. JPCSP to JPCSP is perfectly fine.
> 
> But later also happened on PPSSPP Android vs PC as i remembered.
> 
> It happened on JPCSP vs JPCSP too when playing betwwen old and new version (suspected to be related to VROT/VCOS) https://forums.ppsspp.org/showthread.php?tid=12357&pid=106915#pid106915
> 
> > this error will happens when using jpcsp-rbd345a4 with jpcsp-r3ad1872(or older)
> > as gid15(author of JPCSP) say: That can happen if the game is exchanging values computed using VROT or VCOS through the network: both sides have to use the exact same algorithm.
> > I guess when playing against jpcsp with ppsspp, ppsspp need fixed VROT or VCOS.

Found out that using IR Interperter only on the PC side seem to solve this issue for now and only for GvG.
I only tried 1vs1 Android to PC though 
BTW I remember trying JPCSP vs PSP and it still disconnected there recently...

**Comment by ANR2ME** [comment]
Does using IR interpreter on both Android and PC also worked better?

**Comment by MojoJojoDojo** [comment]
Most Android devices are too slow for using the IR interpreter but I guess it will also work well (also disabling the VFPU stuff in JIT works too).
Still wonder if it's possible to get JPCSP/PPSSPP to work against PSP though...

**Comment by MojoJojoDojo** [comment]
Also it seems like desyncs could happen on real hardware that use just ad-hoc.
Here is an example here for GTA Chinatown Wars when you shoot a taxi lol (happens on PPSSPP too):
https://user-images.githubusercontent.com/33392696/145994817-1268580d-c3bd-4d02-a4c4-a6a78a004f93.mp4


I have a theory that because they ported the game from DS they didnt account for CPU changes between ARM and MIPS but who knows.

**Comment by MojoJojoDojo** [comment]
Lately this issue happens even on PC with localhost lol.

**Comment by Samuel-0116** [comment]
> Hmm.. i'm not sure how to create the cheat properly. This is the code (`li a0,0x2`) which set the FPS to 30 when multiplayer game started, where this `0x00000002` value will be stored at `0x08A78E0C` and being read as arg of `sceDisplayWaitVblankStartMultiCB` ![image](https://user-images.githubusercontent.com/7974720/115826252-cc6ecc00-a434-11eb-880c-124dc9d59d8f.png) After multiplayer ended the game will set the value at `0x08A78E0C` to `1` (60 FPS) using the same function (`z_un_08837a00`)
> 
> I tried to create the cheat like this:
> 
> ```
> _S ULUS-10537
> _G DBZ TTT [US] 
> _C1 60 FPS
> _L 0x08AB7EC8 0x34040001
> _L 0x08A78E0C 0x00000001
> ```
> 
> But when i checked the instruction and the value stored in memory, they're not changing :( did i use the address in the wrong way? (in case the cheat have a base address)
> 
> PS: Replacing the instruction should be sufficient, replacing the vblank value in memory isn't really needed.
> 
> **Update:** Looks like i need to subtract the PSP address with `0x08800000` and added it with `0x20000000` So the correct cheat is:
> 
> ```
> _S ULUS-10537
> _G DBZ TTT [US] 
> _C1 60 FPS
> _L 0x202B7EC8 0x34040001
> ```
> 
> But the animation seems to be moving too fast.. probably not a good idea to use 60 FPS other than for testing purpose.

Sir pls I wanted to speak to you about something can we chat in private

**Comment by Samuel-0116** [comment]
Sir pls I wanted to speak with you about something can we chat in private @

**Comment by MajinDipu** [comment]
> Sir pls I wanted to speak with you about something can we chat in private @

I hope you're talking about @anr2me or @ghost bcz I stopped caring about this game altogether

**Comment by Samuel-0116** [comment]
Yeah I'm talking about @anr2me

**Comment by anr2me** [comment]
i've stopped caring about this game either :) adhoc stuff in general actually

**Comment by MajinDipu** [comment]
> i've stopped caring about this game either :) adhoc stuff in general actually

oh come on! you can't say that! But I would like to take this moment to personally thank you for fixing 99% PSP games' ad-hoc multiplayer.

**Comment by Samuel-0116** [comment]
I wasn't going to ask about ad-hoc

**Comment by Samuel-0116** [comment]
Sir please do you know anything about dbz ttt character roster menu modification

**Comment by hrydgard** [comment]
Hm, weird, and interesting. I guess if you mix devices with different architectures the VFPU stuff could be a desync cause as we are maybe not always 100% identical in the output...  but two devices, both arm64, with the same version of PPSSPP should not desync for this reason.
